### DISCUSSION TITLE
ESM-1v variant effect prediction script for cystein mutants

### DISCUSSION THREAD

**Discussion by konstin** [facebookresearch/esm#422] [Q&A]
I'm trying to predict the effect of substituting amino acids by cystein for smFRET. smFRET requires cysteins to attach fluorophores, but the mutations (and also the fluorophores) shouldn't mess with the secondary structure or the $K_D$. We currently use ConSurf, but my goal is to replace it with something much faster.

I've tried to adapt the [variant-prediction/predict.py](https://github.com/facebookresearch/esm/blob/57da016e5d740a9ac5bcf62c3689a42e88584bc6/examples/variant-prediction/predict.py) example, but unfortunately that's hard to parse for me as I'm not using DMS/MSA input. I've tried to implement the core logic for Masked Marginals and "replace idx with MUTANT_AA" with the following snippet:

```python
wt_encoded = alphabet.get_idx(sequence[idx])
mt_encoded = alphabet.get_idx(MUTANT_AA)

data = [("protein1", sequence)]
batch_labels, batch_strs, batch_tokens = batch_converter(data)
batch_tokens_masked = batch_tokens.clone()
batch_tokens_masked[0, 1 + idx] = alphabet.mask_idx
with torch.no_grad():
    token_probs = torch.log_softmax(
        model(batch_tokens_masked.cuda())["logits"], dim=-1
    )
# add 1 for BOS
mt_prob = token_probs[:, 1 + idx, mt_encoded].item()
wt_prob = token_probs[:, 1 + idx, wt_encoded].item()
score = mt_prob - wt_prob
print(idx, score)
```

Is that the correct way to use ESM-1v? I'm especially uncertain about the indices in `batch_tokens_masked` and `token_probs`, these seem inconsistent to me in `variant-prediction/predict.py`.

I'm using this as part of pretty standalone script, which you might want to add to the examples folder. The script takes a one-line-per-sequence fasta file and outputs all scores in a csv. It also slices sequences > 1022 AA so that they can still be predicted.

<details>
  <summary>Expand</summary>

```python
"""
Variant effect prediction for cystein mutations
"""
# pip install torch==1.12.* fair-esm numpy
from pathlib import Path
from typing import TextIO, Tuple, Generator

import torch
from esm import pretrained, Alphabet

MUTANT_AA = "C"


def slice_sequence(sequence: str) -> Generator[Tuple[str, int, int], None, None]:
    """Since ESM has a length limit of 1024 plus start and end token, we batch into 510 before, our AA, 510 after in the
    middle and the first/last 1021 at the edges.

    Returns sliced sequence, relative index, absolut index"""
    max_len = 1022
    if len(sequence) < max_len:
        # No slicing if the sequence is long enough
        for i in range(len(sequence)):
            yield sequence, i, i
    else:
        for i in range(len(sequence)):
            pad_size = max_len // 2 - 1
            if i <= pad_size:
                yield sequence[: max_len - 1], i, i
            elif len(sequence) - i <= pad_size:
                # assuming (max_len//2-1) == 2
                #  0 1 2 3 4 5 6 7 8 9
                #            ^ ^ ^ * ^
                #            0 1 2 3 4
                # len(sequence) == 4; max_len - 1 == 4
                # len(sequence) - (max_len - 1) == 5
                # i - (len(sequence) - (max_len - 1)) == 3
                yield sequence[-(max_len - 1) :], i - (len(sequence) - (max_len - 1)), i
            else:
                # 0 1 2 3 4 5 6 7 8 9
                #       ^ ^ * ^ ^
                #       0 1 2 3 4
                # So we want [3:8]:
                # left: 5-(max_len//2-1) == 3 which is correct
                # right: 6+(max_len//2-1) == 7, so we add one
                # relative index is (max_len//2-1) == 2
                sliced = sequence[i - pad_size : i + pad_size + 1]
                yield sliced, pad_size, i


def predict_protein(
    pdb_chain: str,
    sequence: str,
    fp: TextIO,
    model,
    alphabet: Alphabet,
    scoring="masked-marginals",
):
    batch_converter = alphabet.get_batch_converter()

    if scoring == "masked-marginals":
        all_token_probs = []

        for (sliced, relative, absolut) in slice_sequence(sequence):
            wt_encoded = alphabet.get_idx(sliced[relative])
            mt_encoded = alphabet.get_idx(MUTANT_AA)

            data = [("protein1", sliced)]
            batch_labels, batch_strs, batch_tokens = batch_converter(data)
            batch_tokens_masked = batch_tokens.clone()
            batch_tokens_masked[0, 1 + relative] = alphabet.mask_idx
            with torch.no_grad():
                token_probs = torch.log_softmax(
                    model(batch_tokens_masked.cuda())["logits"], dim=-1
                )
            # add 1 for BOS
            mt_prob = token_probs[:, 1 + relative, mt_encoded].item()
            wt_prob = token_probs[:, 1 + relative, wt_encoded].item()
            print(pdb_chain, absolut, mt_prob - wt_prob, mt_prob, wt_prob)
            fp.write(f"{pdb_chain},{absolut},{mt_prob},{wt_prob}\n")
            fp.flush()
            all_token_probs.append((absolut, mt_prob, wt_prob))
        assert len(sequence) == len(all_token_probs)
    elif scoring == "wt-marginals":
        # This doesn't support >1024 AA yet
        data = [("protein1", sequence)]
        batch_labels, batch_strs, batch_tokens = batch_converter(data)

        with torch.no_grad():
            token_probs = torch.log_softmax(
                model(batch_tokens.cuda())["logits"], dim=-1
            )

        all_token_probs = []
        for idx in range(len(sequence)):
            wt = sequence[idx]
            assert (
                sequence[idx] == wt
            ), "The listed wildtype does not match the provided sequence"

            wt_encoded = alphabet.get_idx(wt)
            mt_encoded = alphabet.get_idx(MUTANT_AA)

            # add 1 for BOS
            mt_prob = token_probs[:, 1 + idx, mt_encoded].item()
            wt_prob = token_probs[:, 1 + idx, wt_encoded].item()
            print(pdb_chain, idx, mt_prob - wt_prob, mt_prob, wt_prob)
            fp.write(f"{pdb_chain},{idx},{mt_prob},{wt_prob}\n")
            fp.flush()
            all_token_probs.append((idx, mt_prob, wt_prob))
        assert len(sequence) == len(all_token_probs)
    else:
        assert False, scoring


def main():
    lines = Path("sequences_from_pdb_v2.fasta").read_text().splitlines()
    ids = [line[1:] for line in lines[::2]]
    sequences = lines[1::2]

    model_locations = ["esm1v_t33_650M_UR90S_1"]
    # model_locations = ["esm1v_t33_650M_UR90S_1", "esm1v_t33_650M_UR90S_2", "esm1v_t33_650M_UR90S_3", "esm1v_t33_650M_UR90S_4", "esm1v_t33_650M_UR90S_5"]

    model_location = model_locations[0]
    model, alphabet = pretrained.load_model_and_alphabet(model_location)
    model.eval()
    if torch.cuda.is_available():
        model = model.cuda()
        print("Transferred model to GPU")
    else:
        print("No GPU found, this is slow!")

    with Path("scores.csv").open("w") as fp:
        fp.write("id,idx,mt,wt\n")
        fp.flush()
        for pdb_chain, sequence in zip(ids, sequences):
            print(pdb_chain, sequence)
            predict_protein(pdb_chain, sequence, fp, model, alphabet)


if __name__ == "__main__":
    main()


def test_slices_sequences():
    seq = (
        "DEVWVVGMGIVMSLIVLAIVFGNVLVITAIAKFERLQTVTNYFITSLACADLVMGLAVVPFGAAHILMKMWTFGNFWCEFWTSIDVLCVTASIETLCVIAV"
        "DRYFAITSPFKYQSLLTKNKARVIILMVWIVSGLTSFLPIQMHWYRATHQEAINCYAEETCCDFFTNQAYAIASSIVSFYVPLVIMVFVYSRVFQEAKRQL"
        "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
        "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
        "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
        "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
        "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
        "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
        "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
        "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXNIFEMLRIDEGLRLKIYKDTEGYYTIGIGHLLTKSPS"
        "LNAAKSELDKAIGRNTNGVITKDEAEKLFNQDVDAAVRGILRNAKLKPVYDSLDAVRRAALINMVFQMGETGVAGFTNSLRMLQQKRWDEAAVNLAKSRWY"
        "NQTPNRAKRVITTFRTGTWDAY"
    )
    for (sliced, relative, absolut) in slice_sequence(seq):
        assert len(sliced) == 1021, (len(sliced), sliced, relative, absolut)
        assert seq[absolut] == sliced[relative], (sliced, relative, absolut)
    assert (
        "".join(sliced[relative] for (sliced, relative, absolut) in slice_sequence(seq))
        == seq
    )
    assert (
        "".join(seq[absolut] for (sliced, relative, absolut) in slice_sequence(seq))
        == seq
    )
```

</details>
